File: src/main/java/gargoyle/ct/ui/util/render/MessageProviderListCellRenderer.java
package gargoyle.ct.ui.util.render;

import gargoyle.ct.messages.Described;
import gargoyle.ct.messages.MessageProvider;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public final class MessageProviderListCellRenderer<E extends Described> extends FunctionalListCellRenderer<E> {
    public MessageProviderListCellRenderer(final ListCellRenderer originalRenderer,
                                           @NotNull final MessageProvider messageProvider) {
        super(originalRenderer, e -> {
            try {
                return e == null ? "" : messageProvider.getMessage(e.getDescription());
            } catch (RuntimeException e1) {
                return e.getDescription();
            }
        });
    }
}

File: app/src/main/kotlin/gargoyle/ct/ui/util/render/MessageProviderListCellRenderer.kt
package gargoyle.ct.ui.util.render

import gargoyle.ct.util.messages.MessageProvider
import java.util.function.Function
import javax.swing.ListCellRenderer

@Suppress("UNCHECKED_CAST")
class MessageProviderListCellRenderer<E>(
    originalRenderer: ListCellRenderer<*>,
    messageProvider: MessageProvider
) : FunctionalListCellRenderer<E>(originalRenderer as ListCellRenderer<E>, Function { e: E ->
    try {
        messageProvider.getMessage(e.toString())
    } catch (e1: Exception) {
        e.toString()
    }
}) where E : Any
